### PR TITLE
Actor critic adam

### PR DESCRIPTION
**Pull request by ebetica**
Adds a cart pole example and adam

### CODE REVIEW

File: autograd.cpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#include <math.h>
```
Comment (jgehring) [inline autograd.cpp]: Nit: cmath

File: autograd.cpp [hunk 2]
```diff
@@ -310,4 +311,40 @@ void SGD::step() {
   }
 }
 
+void Adam::step() {
+  for (auto pair : model_->parameters()) {
```
Comment (jgehring) [inline autograd.cpp]: Could use `auto& pair` here to not create a copy of the pair.

File: autograd.h [hunk 3]
```diff
@@ -348,4 +348,20 @@ AUTOGRAD_OPTIMIZER_CLASS(SGD) {
   std::unordered_map<std::string, at::Tensor> momentum_buffers;
 };
 
+AUTOGRAD_OPTIMIZER_CLASS(Adam) {
+ public:
+  Adam(Container model, double lr) : Optimizer_CRTP(model), lr_(lr) { }
+  AUTOGRAD_KWARG(Adam, double, momentum, 0, 0);
```
Comment (jgehring) [inline autograd.h]: Adam doesn't have a momentum hyper-parameter, right?

File: test.cpp [hunk 4]
```diff
@@ -14,6 +15,93 @@ using namespace autograd;
 #define CUDA_GUARD std::cerr << "No cuda, skipping test" << std::endl; return
 #endif
 
+class CartPole {
```
Comment (jgehring) [inline test.cpp]: Is this derived from some existing example? If yes, maybe add a short comment pointing to the source.

Comment (ebetica) [reply test.cpp]: Comment was in the step function, but I moved it up here.

**Pull request merged**